File: Data Structures & Algorithms/search-for-word-ii/submission-4.py
class TrieNode:
    def __init__(self):
        self.children={}
        self.ends=False

class Prefix_Tree:
    def __init__(self):
        self.root=TrieNode()
    def add(self,word):
        node=self.root
        for c in word:
            if c not in node.children:
                node.children[c]=TrieNode()
            node=node.children[c]
        node.ends=True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        tri=Prefix_Tree()
        for word in words:
            tri.add(word)
        node=tri.root
        curr=[]
        res=[]
        def dfs(i,j,node):
            if i>=0 and i<len(board) and j>=0 and j<len(board[0]) and board[i][j]!="#" and board[i][j] in node.children:
                t=board[i][j]
                node=node.children[t]
                curr.append(t)
                if node.ends:
                    res.append("".join(curr))
                    node.ends=False
                board[i][j]="#"
                dfs(i+1,j,node)
                dfs(i-1,j,node)
                dfs(i,j+1,node)
                dfs(i,j-1,node)
                board[i][j]=t
                curr.pop()
            else:
                return
        for r in range(len(board)):
            for c in range(len(board[0])):
                dfs(r,c,node)
        return res

File: Data Structures & Algorithms/search-for-word/submission-4.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[0].length;j++){
                if(backtrack(0,board,i,j,word)) return true; 
                }
        }
        return false;
    }
    public static boolean backtrack(int index,char [][]board,int i,int j,String word){
        if(index == word.length()) return true;
        if(i < 0 || i >= board.length || j < 0 || j >= board[0].length || word.charAt(index)!=board[i][j]) return false;
        char c=board[i][j];
        board[i][j] = '#';
        boolean res =backtrack(index+1,board,i-1,j,word) || backtrack(index+1,board,i+1,j,word) || backtrack(index+1,board,i,j-1,word) || backtrack(index+1,board,i,j+1,word);
        board[i][j] = c ;
        return res;
    }

}
